src/cargo/ops/cargo_update.rs

```diff
@@ -207,6 +221,215 @@ pub fn print_lockfile_changes(
+                        name,
+                        version_req
+                    );
+                }
+
+                Ok(dependency)
+            },
+        )?;
+    }
+
+    Ok(upgrades)
+}
+
+/// Update manifests with upgraded versions, and write to disk. Based on cargo-edit.
+pub fn write_manifests(
+    ws: &Workspace<'_>,
+    opts: &UpdateOptions<'_>,
+    upgrades: &HashMap<String, Version>,
+) -> CargoResult<()> {
+    if !opts.breaking {
+        return Ok(());
+    }
+
+    let manifest_paths = std::iter::once(ws.root_manifest())
+        .chain(ws.members().map(|member| member.manifest_path()))
+        .collect::<Vec<_>>();
+
+    for manifest_path in manifest_paths {
+        debug!("writing manifest for `{:?}`", manifest_path);
+
+        let mut local_manifest = LocalManifest::try_new(&manifest_path)?;
+        for dep_table in local_manifest.get_dependency_tables_mut() {
+            for (dep_key, dep_item) in dep_table.iter_mut() {
+                trace!("upgrading dependency {}", dep_key);
+
+                let dep_key = dep_key.get();
+                let dependency = match crate::util::toml_mut::dependency::Dependency::from_toml(
+                    &manifest_path,
+                    dep_key,
+                    dep_item,
+                ) {
+                    Ok(dependency) => dependency,
+                    Err(err) => {
+                        opts.gctx
+                            .shell()
+                            .warn(&format!("ignoring {dep_key}, unsupported entry: {err}"))?;
+                        continue;
+                    }
+                };
+
+                if let crate::util::toml_mut::dependency::MaybeWorkspace::Other(_) =
+                    dependency.source_id(opts.gctx)?
+                {
+                    if let (Some(current), Some(latest)) =
+                        (dependency.version(), upgrades.get(&dependency.name))
+                    {
+                        if let Some(new_req_string) = upgrade_requirement(current, latest)? {
+                            let mut dep = dependency.clone();
+
+                            if let Some(Source::Registry(source)) = dependency.source() {
+                                let mut source = source.clone();
+                                source.version = new_req_string;
+                                dep.source = Some(Source::Registry(source));
+                            }
+
+                            *dep_item = dep.to_toml(manifest_path);
+                        }
+                    }
+                }
+            }
+        }
+
+        if opts.dry_run {
+            opts.gctx.shell().warn(format!(
+                "not upgrading manifest `{}` due to dry run",
+                manifest_path.display()
+            ))?;
+        } else {
+            debug!("writing upgraded manifest to {}", manifest_path.display());
+            local_manifest.write()?;
+        }
```

Comment:
- Should we avoid the `write` / dry-run message if this manifest didn't change?
- Do we get duplicate `write`s / dry-run messages when there is a root package in the workspace?